File: src/resolver_sim/allocation/roots.clj
(ns resolver-sim.allocation.roots
  "Root and digest construction for the allocation kernel.

   All roots are SHA-256 domain-separated hashes over the canonical binary
   encoding (CANONICAL_HASH_SPEC_V1_BINARY_ENCODING_ABI). The Clojure and Rust
   implementations must produce the same exact bytes.

   The result root is a Merkle tree over result leaves using the existing PRF
   Merkle convention (EVIDENCE_COMMITMENT_SPEC_V1):
     leaf  = SHA256(EVIDENCE_MERKLE_LEAF_V1  || leaf_digest)
     node  = SHA256(EVIDENCE_MERKLE_NODE_V1  || left || right)
   An odd level duplicates its final node. Leaf digests are themselves
   domain-separated hashes under :result-root."
  (:require [clojure.string :as str]
            [resolver-sim.hash.canonical :as hc]))

(def ^:private merkle-leaf-tag (.getBytes "EVIDENCE_MERKLE_LEAF_V1" "UTF-8"))
(def ^:private merkle-node-tag (.getBytes "EVIDENCE_MERKLE_NODE_V1" "UTF-8"))

(defn- hex->raw
  "Convert a 0x-prefixed or bare 64-char hex string to raw 32 bytes."
  [hex-str]
  (let [hex (if (str/starts-with? hex-str "0x") (subs hex-str 2) hex-str)]
    (byte-array (map #(Integer/parseInt % 16)
                     (map #(apply str %) (partition 2 hex))))))

(defn- concat-bytes
  [& bas]
  (let [out (byte-array (reduce + (map count bas)))]
    (loop [idx 0, bas bas]
      (if (seq bas)
        (let [ba (first bas)]
          (System/arraycopy ba 0 out idx (count ba))
          (recur (+ idx (count ba)) (rest bas)))
        out))))

(defn merkle-leaf
  "Raw 32-byte Merkle leaf hash: SHA256(EVIDENCE_MERKLE_LEAF_V1 || digest)."
  [^bytes digest]
  (hc/hash-bytes (concat-bytes merkle-leaf-tag digest)))

(defn merkle-node
  "Raw 32-byte internal node hash: SHA256(EVIDENCE_MERKLE_NODE_V1 || left || right)."
  [^bytes left ^bytes right]
  (hc/hash-bytes (concat-bytes merkle-node-tag left right)))

(defn- bytes->bare-hex
  "Lowercase 64-char hex encoding (no 0x prefix), consistent with domain-hash."
  [^bytes ba]
  (apply str (map #(format "%02x" (bit-and (int %) 0xFF)) ba)))

(defn merkle-root
  "Compute the Merkle root over a sequence of raw 32-byte digests using the
   existing PRF convention. An odd level duplicates its final node."
  [raw-digests]
  (loop [level (vec raw-digests)]
    (cond
      (empty? level)
      (throw (ex-info "Empty Merkle tree" {}))

      (= 1 (count level))
      (bytes->bare-hex (first level))

      :else
      (let [level (if (odd? (count level))
                    (conj level (last level))
                    level)
            next-level (mapv (fn [[l r]] (merkle-node l r))
                             (partition 2 level))]
        (recur next-level)))))

(defn claimant-set-root
  "Domain-separated hash of the canonically ordered claimant set."
  [context]
  (hc/domain-hash :claimant-set (:claimants context)))

(defn outcome-set-root
  "Domain-separated hash of the canonically ordered outcome set."
  [context]
  (hc/domain-hash :outcome-set (:outcomes context)))

(defn proposed-rates-root
  "Domain-separated hash of the proposed rates in outcome canonical order."
  [context]
  (hc/domain-hash :proposed-rates
                  (mapv (fn [rate-entry]
                          {:outcome/id (:outcome/id rate-entry)
                           :numerator (get-in rate-entry [:rate :numerator])
                           :denominator (get-in rate-entry [:rate :denominator])})
                        (:proposed-rates context))))

(defn selected-outcome-hash
  "Domain-separated hash of a selected outcome."
  [outcome]
  (hc/domain-hash :selected-outcome outcome))

(defn result-leaf
  "One result Merkle leaf value tree. Commits round/context identity, claim ID,
   beneficiary (economic owner), input amount, input weight, final allocation,
   selected outcome ID, and result status."
  [{:keys [context-hash input-amount input-weight final-allocation
           selected-outcome-id result-status] :as leaf}]
  {:round/context-hash context-hash
   :claim/id (:claim/id leaf)
   :beneficiary (:beneficiary leaf)
   :input-amount input-amount
   :input-weight input-weight
   :final-allocation final-allocation
   :selected-outcome-id selected-outcome-id
   :result-status result-status})

(defn result-leaf-digest
  "Raw 32-byte digest of a result leaf under the :result-root domain."
  [leaf]
  (let [hex (hc/domain-hash :result-root leaf)]
    (hex->raw hex)))

(defn result-merkle-root
  "Merkle root over result leaves in canonical claimant ordering."
  [result-leaves]
  (merkle-root (mapv result-leaf-digest result-leaves)))

(defn certificate-assertions-digest
  "Digest committing to the allocation context hash, the ordered assertion
   results, the selected outcome, the result root, totals, and the kernel
   version. No diagnostics, timestamps, paths, runtime versions, or proof stubs
   are included."
  [{:keys [allocation-context-hash assertions selected-outcome-id
           selected-outcome-index result-root total-allocated residual-capacity
           allocation-kernel-version]}]
  (hc/domain-hash :certificate-assertions
                  {:allocation-context-hash allocation-context-hash
                   :assertions assertions
                   :selected-outcome-id selected-outcome-id
                   :selected-outcome-index selected-outcome-index
                   :result-root result-root
                   :total-allocated total-allocated
                   :residual-capacity residual-capacity
                   :allocation-kernel-version allocation-kernel-version}))

File: coprocessor/core/src/roots.rs
//! Root construction matching the PRF allocation kernel byte-for-byte.

use crate::canonical::{
    bytes_to_hex, domain_hash, domain_hash_raw, hex32_to_bytes, tags, CanonValue,
};
use sha2::{Digest, Sha256};

/// Raw SHA256(EVIDENCE_MERKLE_NODE_V1 || left || right).
fn merkle_node(left: &[u8], right: &[u8]) -> Vec<u8> {
    let mut hasher = Sha256::new();
    hasher.update(tags::MERKLE_NODE.as_bytes());
    hasher.update(left);
    hasher.update(right);
    hasher.finalize().to_vec()
}

/// Compute the Merkle root over raw digests. An odd level duplicates its final
/// node (EVIDENCE_COMMITMENT_SPEC_V1 §9).
pub fn merkle_root(raw_digests: &[Vec<u8>]) -> Result<String, String> {
    if raw_digests.is_empty() {
        return Err("empty Merkle tree".to_string());
    }
    let mut level: Vec<Vec<u8>> = raw_digests.to_vec();
    while level.len() > 1 {
        if level.len() % 2 == 1 {
            let last = level.last().unwrap().clone();
            level.push(last);
        }
        let mut next = Vec::with_capacity(level.len() / 2);
        for pair in level.chunks(2) {
            next.push(merkle_node(&pair[0], &pair[1]));
        }
        level = next;
    }
    Ok(bytes_to_hex(&level[0]))
}

/// The Merkle root of a set of result leaves in canonical claimant order.
pub fn result_merkle_root(leaves: &[CanonValue]) -> Result<String, String> {
    let digests: Vec<Vec<u8>> = leaves
        .iter()
        .map(|leaf| domain_hash_raw(tags::RESULT_ROOT, leaf))
        .collect();
    merkle_root(&digests)
}

/// Claimant-set root: domain hash of the canonically ordered claimant set.
pub fn claimant_set_root(claimants: &CanonValue) -> String {
    domain_hash(tags::CLAIMANT_SET, claimants)
}

/// Outcome-set root: domain hash of the canonically ordered outcome set.
pub fn outcome_set_root(outcomes: &CanonValue) -> String {
    domain_hash(tags::OUTCOME_SET, outcomes)
}

/// Proposed-rates root.
pub fn proposed_rates_root(rates: &CanonValue) -> String {
    domain_hash(tags::PROPOSED_RATES, rates)
}

/// Rate-derived summary hash (wraps the summary in a schema envelope, matching
/// the PRF projection).
pub fn rate_derived_summary_hash(summary: &CanonValue) -> String {
    let envelope = CanonValue::map(vec![
        (
            CanonValue::keyword("schema-version"),
            CanonValue::str("rate-derived-summary.v1"),
        ),
        (CanonValue::keyword("summary"), summary.clone()),
    ]);
    domain_hash(tags::RATE_DERIVED_SUMMARY, &envelope)
}

/// Selected-outcome hash.
pub fn selected_outcome_hash(outcome: &CanonValue) -> String {
    domain_hash(tags::SELECTED_OUTCOME, outcome)
}

/// Certificate assertions digest.
pub fn certificate_assertions_digest(preimage: &CanonValue) -> String {
    domain_hash(tags::CERTIFICATE_ASSERTIONS, preimage)
}

/// CERTIFICATE_ASSERTIONS_V2 digest: extends v1 by committing the round
/// lifecycle observation.
pub fn certificate_assertions_digest_v2(preimage: &CanonValue) -> String {
    domain_hash(tags::CERTIFICATE_ASSERTIONS_V2, preimage)
}

/// Validate a 0x-prefixed or bare 64-char lowercase hex 32-byte hash.
pub fn normalize_hash(s: &str) -> Option<String> {
    let bare = s.strip_prefix("0x").unwrap_or(s);
    if bare.len() != 64
        || !bare
            .bytes()
            .all(|b| b.is_ascii_hexdigit() && (b as char).is_lowercase())
    {
        return None;
    }
    Some(format!("0x{}", bare))
}

pub fn raw_of_hex(s: &str) -> Option<Vec<u8>> {
    hex32_to_bytes(s.strip_prefix("0x").unwrap_or(s))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn odd_level_duplicates_final_node() {
        // single node -> root is that node itself (leaf already applied)
        let d1 = vec![1u8; 32];
        let root1 = merkle_root(std::slice::from_ref(&d1)).unwrap();
        let root2 = merkle_root(std::slice::from_ref(&d1)).unwrap();
        assert_eq!(root1, root2);
    }

    #[test]
    fn different_leaves_differ() {
        let a = vec![1u8; 32];
        let b = vec![2u8; 32];
        let ab = merkle_root(&[a.clone(), b.clone()]).unwrap();
        let ba = merkle_root(&[b.clone(), a.clone()]).unwrap();
        assert_ne!(ab, ba);
    }
}
